beacon_node/http_api/src/attestation_rewards.rs

```diff
@@ -0,0 +1,260 @@
+                warp_utils::reject::custom_not_found(format!("Unable to get base_reward! {:?}", e))
+            })?;
+
+            let reward_numerator = base_reward
+                .safe_mul(weight)
+                .and_then(|reward_numerator| {
+                    reward_numerator.safe_mul(unslashed_participating_increments)
+                })
+                .map_err(|_| {
+                    warp_utils::reject::custom_server_error(
+                        "Unable to calculate reward numerator".to_owned(),
+                    )
+                })?;
+
+            let ideal_reward = reward_numerator
+                .safe_div(active_increments)
+                .and_then(|ideal_reward| ideal_reward.safe_div(WEIGHT_DENOMINATOR))
+                .map_err(|_| {
+                    warp_utils::reject::custom_server_error(
+                        "Unable to calculate ideal_reward".to_owned(),
+                    )
+                })?;
+
+            if !state.is_in_inactivity_leak(previous_epoch, spec) {
+                ideal_rewards_hashmap.insert((*flag_index, effective_balance_eth), ideal_reward);
+            } else {
+                ideal_rewards_hashmap.insert((*flag_index, effective_balance_eth), 0);
+            }
+        }
+    }
+
+    let ideal_rewards: Vec<IdealAttestationRewards> = ideal_rewards_hashmap
+        .iter()
+        .fold(
+            HashMap::new(),
+            |mut acc, ((_flag_index, effective_balance_eth), ideal_reward)| {
+                let entry =
+                    acc.entry(*effective_balance_eth as u32)
+                        .or_insert(IdealAttestationRewards {
+                            effective_balance: *effective_balance_eth as u64,
+                            head: 0,
+                            target: 0,
+                            source: 0,
+                        });
+                match flag_index {
+                    TIMELY_SOURCE_FLAG_INDEX => entry.source += *ideal_reward,
+                    TIMELY_TARGET_FLAG_INDEX => entry.target += *ideal_reward,
+                    TIMELY_HEAD_FLAG_INDEX => entry.head += *ideal_reward,
+                    _ => {}
+                }
+                acc
+            },
+        )
+        .into_values()
+        .collect();
+
+    //--- Calculate total rewards ---//
+    let mut total_rewards = Vec::new();
+
+    let index = participation_cache.eligible_validator_indices();
+
+    for validator_index in index {
+        let eligible = state
+            .is_eligible_validator(previous_epoch, *validator_index)
+            .map_err(|_| {
+                warp_utils::reject::custom_server_error("Unable to get eligible".to_owned())
+            })?;
+
+        let balance = state.get_balance_mut(*validator_index).map_err(|_| {
+            warp_utils::reject::custom_server_error("Unable to get balance".to_owned())
+        })?;
+        *balance = (*balance as f64).round().clamp(0.0, 32.0) as u64;
```

Comment:
We should look up the effective balance (in gwei) from the state here rather than trying to infer it from the `validator.balance`. The way that effective balance is updated is a bit nuanced, and we can avoid the complexity by just looking it up instead:

```suggestion
        let effective_balance = state.get_effective_balance(*validator_index)?;
        let effective_balance_eth = effective_balance.safe_div(spec.effective_balance_increment)?;
```

We then divide by a billion (i.e. `spec.effective_balance_increment`) to get the effective balance in ETH.